services/database/scripts/migrateUserStates.js

```diff
@@ -0,0 +1,210 @@
+ */
+async function gatherLoginDatesFromCognito () {
+  const cognitoClient = new CognitoIdentityProviderClient();
+  const UserPoolId = COGNITO_USER_POOL_ID;
+  const pages = paginateListUsers({ client: cognitoClient }, { UserPoolId });
+  const modifyDates = new Map();
+  for await (let page of pages) {
+    for (let user of page.Users ?? []) {
+      const sub = user.Attributes.find(attr => attr.Name === "sub").Value;
+      modifyDates.set(sub, user.UserLastModifiedDate);
+    }
+  }
+  return modifyDates;
+}
+
+/**
+ * Given a user object, ensure it has the correct shape.
+ * If not, modify it in-place.
+ * @param {object} user The AuthUser record
+ * @param {Map<string, string>} loginDates A trustworthy set of login dates
+ * @returns `true` if a change was made; `false` if no change was needed.
+ */
+function updateUserState (user, loginDates) {
+  const { role, states, usernameSub } = user;
+
+  if (states === undefined) {
+    // This user has already been migrated; no update necessary.
+    return false;
+  }
+
+  if (role === "state" && Array.isArray(states)) {
+    user.state = user.states[0];
+  }
+
+  // If we have a more trustworthy date, use it. Otherwise, leave what's there.
+  user.lastLogin = loginDates.get(usernameSub) ?? user.lastLogin;
+
+  delete user.states;
+  delete user.isSuperUser;
+  delete user.password;
+  return true;
+}
+
+/**
+ * Create an iterator for the entire table.
+ * @param {string} TableName 
+ */
+async function * scanAllUsers (TableName) {
+  console.log(`${logPrefix()}Scanning...`);
+  let pageNumber = 0;
+  let totalItemCount = 0;
+  for await (let page of paginateScan({ client }, { TableName })) {
+    pageNumber += 1;
+    let pageItemCount = 0;
+    for (let user of (page.Items ?? [])) {
+      pageItemCount += 1;
+      totalItemCount += 1;
+      yield user;
+    }
+    console.log(`${logPrefix()}Completed scan of page ${pageNumber}; ${pageItemCount} items processed.`);
+  }
+  console.log(`${logPrefix()}Scan complete; ${totalItemCount} items in all.`);
+}
+
+/**
+ * @param {AsyncGenerator<object>} userIterable 
+ */
+async function * filterAndModifyUsers (userIterable, loginDates) {
+  for await (let user of userIterable) {
+    const needsUpdate = updateUserState(user, loginDates);
+    if (needsUpdate) {
+      yield user;
+    }
+  }
+}
+
+/**
+ * @param {AsyncGenerator<object>} itemIterable
+ */
+async function * createBatches (itemIterable) {
+  let batch = [];
+
+  for await (let item of itemIterable) {
+    batch.push(item);
+    if (batch.length === 25) {
+      yield batch;
+      batch = [];
+    }
+  }
+
+  if (batch.length > 0) {
+    yield batch;
+  }
+}
+
+/**
+ * Send BatchWriteCommands to Put the given objects in the given table.
+ * @param {AsyncGenerator<object[]>} batches
+ * @param {string} tableName
+ * @param {(object) => string} idSelector For logging purposes.
+ */
+async function sendBatches (batches, tableName, idSelector) {
+  for (let batch of batches) {
```

Comment:
I believe that loops like this can have `const batch` instead of let because the variable isn't being reused like in `for (let i=0; i<4; i++)`

Reply:
that's what I try to do, I don't know how high of priority that would be though.